src/sync_key_gen.rs

```diff
@@ -279,76 +335,86 @@ impl<NodeUid: Ord + Clone + Debug> SyncKeyGen<NodeUid> {
+    /// Returns the number of complete parts. If this is at least `threshold + 1`, the keys can
     /// be generated, but it is possible to wait for more to increase security.
     pub fn count_complete(&self) -> usize {
-        self.proposals
+        self.parts
             .values()
-            .filter(|proposal| proposal.is_complete(self.threshold))
+            .filter(|part| part.is_complete(self.threshold))
             .count()
     }
 
-    /// Returns `true` if the proposal of the given node is complete.
+    /// Returns `true` if the part of the given node is complete.
     pub fn is_node_ready(&self, proposer_id: &NodeUid) -> bool {
         self.node_index(proposer_id)
-            .and_then(|proposer_idx| self.proposals.get(&proposer_idx))
-            .map_or(false, |proposal| proposal.is_complete(self.threshold))
+            .and_then(|proposer_idx| self.parts.get(&proposer_idx))
+            .map_or(false, |part| part.is_complete(self.threshold))
     }
 
-    /// Returns `true` if enough proposals are complete to safely generate the new key.
+    /// Returns `true` if enough parts are complete to safely generate the new key.
     pub fn is_ready(&self) -> bool {
         self.count_complete() > self.threshold
     }
 
-    /// Returns the new secret key and the public key set.
+    /// Returns the new secret key share and the public key set.
     ///
     /// These are only secure if `is_ready` returned `true`. Otherwise it is not guaranteed that
     /// none of the nodes knows the secret master key.
+    ///
+    /// If we are only an observer node, no secret key share is returned.
     pub fn generate(&self) -> (PublicKeySet, Option<SecretKeyShare>) {
         let mut pk_commit = Poly::zero().commitment();
         let mut opt_sk_val = self.our_idx.map(|_| Fr::zero());
-        let is_complete = |proposal: &&ProposalState| proposal.is_complete(self.threshold);
-        for proposal in self.proposals.values().filter(is_complete) {
-            pk_commit += proposal.commit.row(0);
+        let is_complete = |part: &&ProposalState| part.is_complete(self.threshold);
+        for part in self.parts.values().filter(is_complete) {
+            pk_commit += part.commit.row(0);
             if let Some(sk_val) = opt_sk_val.as_mut() {
-                let row: Poly = Poly::interpolate(proposal.values.iter().take(self.threshold + 1));
+                let row: Poly = Poly::interpolate(part.values.iter().take(self.threshold + 1));
                 sk_val.add_assign(&row.evaluate(0));
             }
         }
         let opt_sk = opt_sk_val.map(SecretKeyShare::from_value);
         (pk_commit.into(), opt_sk)
     }
 
-    /// Handles an `Accept` message or returns an error string.
-    fn handle_accept_or_err(
+    /// Consumes the instance, generates the key set and returns a new `NetworkInfo` with the new
+    /// keys.
+    pub fn into_network_info(self) -> NetworkInfo<NodeUid> {
```

Comment:
We could `impl<NodeUid> From<SyncKeyGen> for NetworkInfo<NodeUid>` here.  I like using conversion traits, but it may be clearer as currently written.

Reply:
The reason why I was hesitating is that maybe `From` suggests that it's a "simple" and cheap conversion, where in reality it does quite a bit of computation to generate the keys, and just happens to consume the `SyncKeyGen` to avoid some cloning.

But on the one hand I'm not sure that `into_network_info` doesn't suggest the same thing; and on the other hand I don't know whether it's only me who would read it that way.

Another option would be renaming it to something like `generate_network_info`.
Not sure what's best?

Reply:
Also, I'd probably assume that if it's `Into`/`From`, it's _always_ fine to do the conversion; but it only produces a safe result after it ate enough `Part` and `Ack` messages.

Reply:
I definitely agree with your point that `Into/From` would infer that the conversion is always possible, which rules the conversion trait out to me.